### PR TITLE
fix: add commitOp to KeyInfo and refactor tests

### LINKED ISSUE #833: KeyInfo should include description and commitOp
### Is your feature request related to a problem? Please describe.

The KeyInfo currently gives information about the key, syncDirection, and conflictInfo. 

### Describe the solution you'd like

KeyInfo should include a description saying about the cause. For example, "Key is an invalid/Bad key ", "Delete of key is successful".
Also, KeyInfo should include the commitOp of the key, Whether it is an UPDATE, DELETE, or UPDATE_META operation.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sitaram-kalluri**
**- What I did**
* Add CommitOperation to the KeyInfo that describes the commit operation of the key upon Sync.
* Deprecate the OnDone callback in SyncService
* Refactor the "atclient_sync_callback_test.dart" in the at_functional_tests to use "expectAsync1" instead of adding delays to assert on the SyncProgress.

**- How I did it**
* Add CommitOperation as an instance variable in the KeyInfo class. Upon sync to remote (or) sync to local, populate the CommitOperation that describes the commit operation performed on the key
* In the "atclient_sync_callback_test.dart" functional test, currently we have a delay of 10 seconds for the sync to complete and then to assert the data in the SyncProgress. After 10 Seconds, the sync listener is removed from listening to the sync progress. This is causing the validations to skip. Hence removed 10 seconds delay and added "expectAsync1" which waits until the "onSyncProgressEvent" is triggered once. 
* Introduce a new class -"FunctionalTestSyncService" in at_functional_tests which when invoked run sync and awaits for sync completion
* Removed the usage of onDone callback in at_end2end_tests and at_functional_tests and replaced with SyncProgressListener.
* Updated the tests in sync_new_tests.dart to use SyncProgressListener instead of onDone callback to await for sync completion.
 
**- How to verify it**
* The unit and functional tests to assert different commit operations when sync happens between client and server.

**- Description for the changelog**
* Add CommitOperation (CommitOp.) to the KeyInfo to describe if the key is created/deleted upon sync

### CODE REVIEW

File: packages/at_client/lib/src/service/sync_service.dart [hunk 1]
```diff
@@ -35,6 +35,7 @@ abstract class SyncService {
   /// Call this method to set the Global onDone callback.
   /// This method will be called when a sync is completed.
   /// When a specific onDone function is passed to the sync Function, Then the specific onDone is called.
+  @Deprecated('Use SyncProgressListener')
```
Comment (gkc) [inline packages/at_client/lib/src/service/sync_service.dart:38]: Shouldn't we also deprecate the onDone parameter of the sync function?

**Review by gkc — CHANGES_REQUESTED** [review]

File: packages/at_client/lib/src/service/sync_service.dart (hunk 1 above)
Comment (sitaram-kalluri) [reply packages/at_client/lib/src/service/sync_service.dart:38]: @gkc: The "onDone" callback is used sync_new_test.dart (sync unit tests) and deprecating in the sync method raised below caution when running the dart analyzer (there are 15 such issues reported by dart analyzer).
`
info • test/sync_new_test.dart:2577:24 • 'onDone' is deprecated and shouldn't be used. Use SyncProgressListener. Try replacing the use of the deprecated
          member with the replacement. • deprecated_member_use_from_same_package
`
Shall we refactor the tests to use SyncProgressListener and then deprecate the "onDone" in the sync method?

Comment (gkc) [reply packages/at_client/lib/src/service/sync_service.dart:38]: Yes

Please also check all existing app source code in atsign-foundation for uses of onDone and raise tickets to change that code

Comment (sitaram-kalluri) [reply packages/at_client/lib/src/service/sync_service.dart:38]: Sure Gary, Can the deprecating of "onDone" callback and refactoring of tests go in a separate PR?

Comment (gkc) [reply packages/at_client/lib/src/service/sync_service.dart:38]: Yes, if you like? But why not do it now when you're already working in that area of the code?

If you do go ahead to merge this PR as-is, please first check all existing app source code in atsign-foundation for uses of onDone and raise tickets to change that code

Comment (sitaram-kalluri) [reply packages/at_client/lib/src/service/sync_service.dart:38]: Refactored the tests to use SyncProgressListener and removed the usage of onDone callback.

**Comment by gkc** [comment]
@sitaram-kalluri can you merge trunk into this branch and resolve the conflicts please? I'd like to review this and, all being well, get it merged tomorrow. Thanks!

**Comment by sitaram-kalluri** [comment]
> @sitaram-kalluri can you merge trunk into this branch and resolve the conflicts please? I'd like to review this and, all being well, get it merged tomorrow. Thanks!

@gkc: Resolved the merge conflicts.

File: packages/at_client/lib/src/util/at_client_util.dart [hunk 2]
```diff
@@ -146,3 +147,18 @@ extension ParseEncodingTypeToString on EncodingType {
...
+    case '+':
+      return CommitOp.UPDATE;
+    case '#':
+      return CommitOp.UPDATE_META;
+    case '*':
+      return CommitOp.UPDATE_ALL;
+    case '-':
+      return CommitOp.DELETE;
+    default:
+      return CommitOp.UPDATE;
```
Comment (gkc) [inline packages/at_client/lib/src/util/at_client_util.dart]: Default case should be to throw an exception.

File: packages/at_client/lib/src/service/sync_service_impl.dart [hunk 3]
```diff
@@ -114,7 +114,9 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
   }
 
   @override
-  void sync({Function? onDone, Function? onError}) {
+  void sync(
+      {@Deprecated('Use SyncProgressListener') Function? onDone,
```
Comment (gkc) [inline packages/at_client/lib/src/service/sync_service_impl.dart:118]: Please make sure the deprecated annotations are in both sync_service.dart and sync_service_impl.dart

**Review by gkc — CHANGES_REQUESTED** [review]

File: packages/at_client/lib/src/util/at_client_util.dart (hunk 2 above)
Comment (sitaram-kalluri) [reply packages/at_client/lib/src/util/at_client_util.dart]: Modified code to throw exception from default case.

File: packages/at_client/lib/src/service/sync_service_impl.dart (hunk 3 above)
Comment (sitaram-kalluri) [reply packages/at_client/lib/src/service/sync_service_impl.dart:118]: Added deprecated tag to onDone callback in sync_service.dart as well.

**Comment by gkc** [comment]
@sitaram-kalluri I know it's unrelated to your PR, but since an end2end test (bypasscache_test.dart) is frequently failing, please can you figure out whatever timing issue is causing it to fail, and fix it, as part of this PR?

**Review by gkc — APPROVED** [review]
LGTM

**Comment by gkc** [comment]
> @sitaram-kalluri I know it's unrelated to your PR, but since an end2end test (bypasscache_test.dart) is frequently failing, please can you figure out whatever timing issue is causing it to fail, and fix it, as part of this PR?

Actually, ignore my previous request. Please fix that issue in a separate PR, assuming we can get the tests to eventually pass for this PR by re-running the checks. (I'm re-re-running the checks now)

**Comment by sitaram-kalluri** [comment]
> > @sitaram-kalluri I know it's unrelated to your PR, but since an end2end test (bypasscache_test.dart) is frequently failing, please can you figure out whatever timing issue is causing it to fail, and fix it, as part of this PR?
> 
> Actually, ignore my previous request. Please fix that issue in a separate PR, assuming we can get the tests to eventually pass for this PR by re-running the checks. (I'm re-re-running the checks now)

Sure Gary, i will look at the bypasscache_test and fix it.

**Pull request merged**